training/creating_dataset.py

```diff
@@ -15,117 +13,92 @@
 
-    def inner_move_values_level_up(df: pd.DataFrame) -> pd.DataFrame:
-        return df.applymap(lambda x: x.get(value_name), na_action="ignore")
-
-    return inner_move_values_level_up
-
-
-def get_subcolumn(book: pd.DataFrame, subcolumn_path: str) -> pd.DataFrame:
-    """
-    Gets subcolumn of given DataFrame according to given path
-    :param book: DataFrame with all data from book(s)
-    :param subcolumn_path: path to subcolumn
-    :return: chosen subcolumn as DataFrame
+    :param paths_to_books: list of paths of books to load
+    :param characteristics: list of characteristics to load
+    :return: DataFrame with monsters (NPC) from chosen books and with chosen characteristics and their origin book
     """
-    subcol = deepcopy(book)
-    for col in subcolumn_path.split("/"):
-        subcol = pd.DataFrame(data=unpack_column(subcol, col))
+    data = []
+
+    for path in paths_to_books:
+        is_path_correct(path)
+        with open(path) as file:
+            # loading json strings from files
+            data += [json.loads(line) for line in file]
+            file.close()
 
-    return subcol
+    bestiary = pd.json_normalize(data)
+    # only npc monsters
+    bestiary = bestiary[bestiary["type"] == "npc"]
+    # only system column (all characteristics are there)
+    bestiary = bestiary.filter(regex="system", axis="columns")
 
+    COLS_TO_EXTRACT = pd.DataFrame(
+        data=[
+            (characteristic, CHARACTERISTICS_COLUMNS.get(characteristic))
+            for characteristic in characteristics + ["book", "level"]
+        ],
+        columns=["target_name", "raw_name"],
+    )
 
-def load_subcolumn_as_value(
-    column_name: str, original_column_name: str = "value"
-) -> callable:
-    """
-    Returns a function that creates DataFrame with chosen value_name of given DataFrame
-    and changes column name to chosen one
-    :param column_name: name that should be given to result dataframe column
-    :param original_column_name: name of subcolumn which value will be returned
-    :return: function to create Dataframe with chosen column name and values from a chosen subcolumn of current column
-    """
+    raw_names = COLS_TO_EXTRACT["raw_name"]
+    target_names = COLS_TO_EXTRACT["target_name"]
 
-    def subcolumn_as_value(df: pd.DataFrame) -> pd.DataFrame:
-        result_df = pd.DataFrame(data=df[original_column_name])
-        result_df.columns = [column_name]
-        return result_df
+    # to not have Series names as a part of final df
+    target_names.name = None
 
-    return subcolumn_as_value
+    df = bestiary[raw_names]
+    df.columns = target_names
 
+    if "focus" in df.columns:
+        # silent warning (SettingWithCopyWarning) about view and copy
+        # we don't need to go back to the original df - no matter if it is a view
+        with pd.option_context("mode.chained_assignment", None):
+            df["focus"] = df["focus"].fillna(-1)
```

Comment:
Wartością domyślną focusu nie jest 0?

Reply:
Istnieją potwory, które od startu mają focus = 0, więc to był sposób na odróżnienie od  nich tych, które mają wartości brakujące

Reply:
To jest jak najbardziej ok, żeby tego nie rozróżniać. Tzn. jeżeli coś nie ma focusu, no to go nie ma. Zrobienie imputacji wartości brakujących z pomocą wiedzy domenowej jest zdecydowanie git. Dałbym zero w takiej sytuacji.

BTW - jeżeli użyjecie LightGBM jako regresora w przyszłości, to on przyjmuje tak po prostu wartości brakujące, i nawet potrafi tego inteligentnie użyć